CMakeLists.txt

```diff
@@ -0,0 +1,294 @@
+cmake_minimum_required(VERSION 3.14 FATAL_ERROR)
+
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+project(bladebit)
+
+set(CMAKE_CONFIGURATION_TYPES release debug)
+
+if(NOT CMAKE_BUILD_TYPE)
+    set(CMAKE_BUILD_TYPE "release"
+        CACHE STRING "Possible values are: release, debug"
+        FORCE
+   )
+endif()
+message("Config: ${CMAKE_BUILD_TYPE}")
+
+
+# Grab BLS
+include(FetchContent)
+
+set(BUILD_BLS_PYTHON_BINDINGS "0" CACHE STRING "0")
+set(BUILD_BLS_TESTS "0" CACHE STRING "")
+set(BUILD_BLS_BENCHMARKS "0" CACHE STRING "")
+
+FetchContent_Declare(
+  bls
+  GIT_REPOSITORY https://github.com/Chia-Network/bls-signatures.git
+  GIT_TAG        1.0.6
+)
+FetchContent_MakeAvailable(bls)
+
+set(CMAKE_C_FLAGS_RELEASE_INIT)
+set(CMAKE_C_FLAGS_DEBUG_INIT)
+set(CMAKE_CXX_FLAGS_RELEASE_INIT)
+set(CMAKE_CXX_FLAGS_DEBUG_INIT)
+
+
+set(c_opts)
+set(link_opts)
+# set(cxx_opts)
+
+set(release_c_opts)
+set(debug_c_opts)
+set(dev_c_opts)
+
+set(release_link_opts)
+set(debug_link_opts)
+
+# Config
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
+
+    # MSVC
+    set(c_opts
+        /std:c++17
+        /Zc:__cplusplus
+        /MP
+        /Zi
+        /EHsc-
+        # /Wall
+        /W3
+        /WX
+        /FIpch.h 
+        /wd4068
+        /wd4464
+        /wd4668
+        /wd4820
+        /wd4514
+        /wd4626
+        /wd5027
+        /DUNICODE=1
+        /DWIN32_LEAN_AND_MEAN=1
+        /DNOMINMAX=1
+        /D_CRT_SECURE_NO_WARNINGS=1
+        /D_HAS_EXCEPTIONS=0
+        ${c_opts})
+    
+    set(link_opts 
+        /SUBSYSTEM:CONSOLE
+        ${link_opts})
+
+    set(release_c_opts 
+        /MT
+        /O2
+        /Oi
+        /Gy
+        /GL
+        /DNDEBUG=1
+        /DNDEBUG=1
+        ${release_c_opts})
+
+    set(debug_c_opts 
+        /MTd
+        /Od
+        ${debug_c_opts})
+
+    set(dev_c_opts 
+        ${dev_c_opts})
+
+    
+    set(release_link_opts 
+        /DEBUG:FULL
+        /OPT:REF,IFC,LBR
+        ${release_link_opts})
+
+    set(debug_link_opts
+        /DEBUG:FASTLINK
+        /OPT:NOREF,NOICF,NOLBR
+        /INCREMENTAL
+        ${debug_link_opts})
+
+    # Dependency config
+    target_compile_options(bls     PRIVATE /MP)
+    target_compile_options(relic_s PRIVATE /MP)
+    target_compile_options(sodium  PRIVATE /MP)
+
+    target_compile_options(bls     PRIVATE $<$<CONFIG:release>:/MT>)
+    target_compile_options(relic_s PRIVATE $<$<CONFIG:release>:/MT>)
+    target_compile_options(sodium  PRIVATE $<$<CONFIG:release>:/MT>)
+
+    target_compile_options(bls     PRIVATE $<$<CONFIG:debug>:/MTd>)
+    target_compile_options(relic_s PRIVATE $<$<CONFIG:debug>:/MTd>)
+    target_compile_options(sodium  PRIVATE $<$<CONFIG:debug>:/MTd>)
+    
+else()
+
+    # *Nix
+    set(CMAKE_C_FLAGS --include=pch.h -Wall -Wno-comment -g ${CMAKE_C_FLAGS})
+    set(CMAKE_CXX_FLAGS -std=c++17 ${CMAKE_CXX_FLAGS})
```

Comment:
Maybe target compile options instead of altering CMAKE_C_FLAGS and co directly?

Same idea as above about setting the std flag btw, would idiomatically be handled through CMAKE_CXX_STANDARD.
Also same idea as above about PCH handling, it can idiomatically be handled through `target_precompile_headers`